pkg/service/model_build_load_balancer.go

```diff
@@ -109,25 +111,66 @@ func (t *defaultModelBuildTask) buildLoadBalancerTags(ctx context.Context) (map[
+func (t *defaultModelBuildTask) buildLoadBalancerSubnetMappings(ctx context.Context, scheme elbv2model.LoadBalancerScheme, ec2Subnets []*ec2.Subnet) ([]elbv2model.SubnetMapping, error) {
 	var eipAllocation []string
 	eipConfigured := t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixEIPAllocations, &eipAllocation, t.service.Annotations)
-	if eipConfigured && len(eipAllocation) != len(ec2Subnets) {
-		return []elbv2model.SubnetMapping{}, errors.Errorf("number of EIP allocations (%d) and subnets (%d) must match", len(eipAllocation), len(ec2Subnets))
+	var privateIpv4Addresses []string
+	ipv4Configured := t.annotationParser.ParseStringSliceAnnotation(annotations.SvcLBSuffixPrivateIpv4Addresses, &privateIpv4Addresses, t.service.Annotations)
+
+	// Validation
+	if eipConfigured && ipv4Configured {
+		return []elbv2model.SubnetMapping{}, errors.Errorf("only one of EIP allocations or PrivateIpv4Addresses can be set")
+	}
+	if eipConfigured {
+		if scheme == elbv2model.LoadBalancerSchemeInternal {
+			return []elbv2model.SubnetMapping{}, errors.Errorf("EIP allocations can only be set for internet facing load balancers")
+		} else if len(eipAllocation) != len(ec2Subnets) {
+			return []elbv2model.SubnetMapping{}, errors.Errorf("number of EIP allocations (%d) and subnets (%d) must match", len(eipAllocation), len(ec2Subnets))
+		}
+	}
+	if ipv4Configured {
+		if scheme == elbv2model.LoadBalancerSchemeInternetFacing {
+			return []elbv2model.SubnetMapping{}, errors.Errorf("PrivateIpv4Addresses can only be set for internal balancers")
+		} else if len(privateIpv4Addresses) != len(ec2Subnets) {
+			return []elbv2model.SubnetMapping{}, errors.Errorf("number of PrivateIpv4Addresses (%d) and subnets (%d) must match", len(privateIpv4Addresses), len(ec2Subnets))
+		}
 	}
+
 	subnetMappings := make([]elbv2model.SubnetMapping, 0, len(ec2Subnets))
 	for idx, subnet := range ec2Subnets {
 		mapping := elbv2model.SubnetMapping{
 			SubnetID: aws.StringValue(subnet.SubnetId),
 		}
-		if idx < len(eipAllocation) {
+		if eipConfigured {
 			mapping.AllocationID = aws.String(eipAllocation[idx])
 		}
+		if ipv4Configured {
+			ip, err := t.getMatchingIPforSubnet(ctx, subnet, privateIpv4Addresses)
+			if err != nil {
+				return []elbv2model.SubnetMapping{}, err
+			}
+			mapping.PrivateIPv4Address = aws.String(ip)
+		}
 		subnetMappings = append(subnetMappings, mapping)
 	}
 	return subnetMappings, nil
 }
 
+// Return the ip address which is in the subnet. Error if not match
+// Can be extended for ipv6 if required
+func (t *defaultModelBuildTask) getMatchingIPforSubnet(_ context.Context, subnet *ec2.Subnet, privateIpv4Addresses []string) (string, error) {
+	_, ipv4Net, err := net.ParseCIDR(*subnet.CidrBlock)
+	if err != nil {
+		return "", errors.Wrap(err, "subnet CIDR block could not be parsed")
+	}
+	for _, ipString := range privateIpv4Addresses {
+		if ipv4Net.Contains(net.ParseIP(ipString)) {
```

Comment:
i think it's better to explicitly check for `net.ParseIP(ipString)` returns nil or not.
currently it replies on a implementation detail that `ipv4Net.Contains(nil)` didn't crash and returns false, which isn't obvious from the code.

also, we should have a test case where the IP is not a valid IPv4Address. (since it's user-input)

Reply:
Added, thanks for the comment.

I have added some uncommitted `fmt` changes too. I didn't commit the changes to the CRD yaml (like in my other PR #1785). I think it is a difference in version when I run `make ...`